src/pykeen/triples/splitting.py

```diff
@@ -0,0 +1,314 @@
+        if actual_size != exp_size:
+            logger.warning(
+                f'Requested ratio[{i}]={exp_ratio:.3f} (equal to size {exp_size}), but got {actual_ratio:.3f} '
+                f'(equal to size {actual_size}) to ensure that all entities/relations occur in train.',
+            )
+
+    return triples_groups
+
+
+def _split_triples_with_train_coverage(
+    triples: np.ndarray,
+    sizes: Sequence[int],
+) -> Sequence[np.ndarray]:
+    """
+    Split triples into groups ensuring that all entities and relations occur in the first group of triples.
+
+    :param triples: shape: (num_triples, 3)
+        The triples.
+    :param sizes:
+        The group sizes.
+
+    :return:
+        The groups, where the first group is guaranteed to contain each entity and relation at least once.
+    """
+    seed_mask = _get_cover_randomized_greedy(triples)
+    train_seed = triples[seed_mask]
+    remaining_triples = triples[~seed_mask]
+    # TODO: what to do if train_seed.shape[0] > sizes[0]
+    remaining_sizes = (sizes[0] - train_seed.shape[0],) + tuple(sizes[1:])
+    train, *rest = _split_triples(remaining_triples, remaining_sizes)
+    return (np.concatenate([train_seed, train]), *rest)
+
+
+def _get_cover_deterministic(triples: np.ndarray) -> np.ndarray:
+    """
+    Get a coverage mask for all entities and relations.
+
+    :param triples: shape: (n, 3)
+        The triples.
+
+    :return: shape: (n,)
+        A boolean mask indicating whether the triple is part of the cover.
+    """
+    num_entities = triples[:, [0, 2]].max() + 1
+    num_relations = triples[:, 1].max() + 1
+    num_triples = triples.shape[0]
+
+    # index
+    entities = numpy.full(shape=(num_entities,), fill_value=-1, dtype=numpy.int64)
+    relations = numpy.full(shape=(num_relations,), fill_value=-1, dtype=numpy.int64)
+    h, r, t = triples.T
+    triple_id = numpy.arange(num_triples)
+    entities[h] = relations[r] = entities[t] = triple_id
+
+    if entities.min() < 0:
+        raise RuntimeError(f'unfilled entities exist: {entities}')
+    if relations.min() < 0:
+        raise RuntimeError(f'unfilled relations exist: {relations}')
+
+    # select
+    seed_mask = numpy.zeros(shape=(num_triples,), dtype=numpy.bool)
+    seed_mask[np.r_[entities, relations]] = True
+    return seed_mask
+
+
+def _select_to_cover(
+    index: Mapping[int, Sequence[int]],
+    covered: Set[int],
+    covered_relations: Set[int],
+    covered_entities: Set[int],
+    triples: np.ndarray,
+    seed_mask: np.ndarray,
+) -> None:
+    for i, triple_ids in index.items():
+        if i in covered:
+            continue
+        # randomly select triple
+        tr_id = random.choice(triple_ids)
+        seed_mask[tr_id] = True
+        # update coverage
+        h_id, r_id, t_id = triples[tr_id]
+        covered_entities.update(h_id, t_id)
+        covered_relations.add(r_id)
+
+
+def _get_cover_randomized_greedy(triples):
```

Comment:
Is this method still used? does it offer any advantages?

Otherwise I would opt for removing it again.

Reply:
I didn't even benchmark this one. What would the advantages be versus the other new method? Is it faster?

Reply:
It was there first :wink: I do not think there is an advantage. It is expected to be slower.